File: cwms_read/cwms_read.py
# -*- coding: utf-8 -*-

import requests
import json
import pandas as pd
from datetime import datetime, timedelta
import sys



def fill_index(df, start_date, end_date, freq):
    start = datetime(*start_date)
    end = datetime(*end_date)
    end = end.replace(hour = 23, minute = 0, second = 0)
    date = pd.date_range(start, end, freq = freq)
    date = [pd.Timestamp(x) for x in date]
    df = df.reindex(date)
    df.index.rename('date', inplace = True)
    return df


def get_frequency(path: str)->str:
    freq = path.split('.')[3]
    if '~' in freq: return False
    elif 'Hour' in freq:
        return freq.split('Hour')[0] + 'H'
    elif 'Day' in freq:
        return freq.split('Day')[0] + 'D'
    elif 'Min' in freq:
        return freq.split('Min')[0] + 'min'
    return False
    
def time_window_url(paths, public=True, lookback = 7, start_date = False, end_date = False, timezone = 'PST'):
    """
    helper function for cwms_read
    
    Arguments:  
        
        path -- cwms data path, 
        public -- boolean, 
        start_date -- date integer tuple format (YYYY, m, d)
        end_date -- date integer tuple format (YYYY, m, d)
        timezone -- optional keyword argument if time zone is specified.  
                    Defaults to 'PST' if nothing set
    Returns:
        
        url -- url string of CWMS data webservice for the specified 
               data path and time window
               
    """

    if isinstance(paths, list): 
        path = '%22%2C%22'.join(paths)
    else: path = paths
        
    if public:
        url = r'http://pweb.crohms.org/dd/common/web_service/webexec/getjson?timezone=TIMEZONE_&query=%5B%22PATH%22%5D&'
    else:
        url = r'http://nwp-wmlocal2.nwp.usace.army.mil/common/web_service/webexec/getjson?timezone=TIMEZONE_&query=%5B%22PATH%22%5D&'
    
    url = url.replace('PATH', path).replace('TIMEZONE_', timezone)
    if lookback:
        time = 'backward=' + str(lookback) + 'd'
        url = url + time
    else:
        url = url + 'startdate=START_MONTH%2FSTART_DAY%2FSTART_YEAR+00%3A00&enddate=END_MONTH%2FEND_DAY%2FEND_YEAR+23%3A00'
        sy,sm,sd = start_date
        start_date = datetime(sy,sm,sd)
        ey,em,ed = end_date
        end_date = datetime(ey,em,ed)
        url = url.replace('START_MONTH', str(start_date.month)).replace('START_DAY', str(start_date.day)).replace('START_YEAR', str(start_date.year))
        url = url.replace('END_MONTH', str(end_date.month)).replace('END_DAY', str(end_date.day)).replace('END_YEAR', str(end_date.year))
    
    return url

    

def get_cwms(paths, col_names = None, public = True, fill = True, **kwargs):
    
    """
    A function to parse CWMS json data from webservice into a pandas dataframe
        
    Positional Arguments: 
        paths -- single string or list of string of CWMS data paths, example: 'TDDO.Temp-Water.Inst.1Hour.0.GOES-REV' 
    Keyword Arguments:
        
        The web service can either get a lookback, which is just a number of 
        days from the current day, or a time window.  Two key word arguments are 
        needed for a time wondow, start_date, end_date.  The Timezone can also
        be set.
        
        lookback    --  The number of days from current day to grab data.
                        (int or str) 
                        example: 7
                        
        start_date  --  The start of a time window (tuple) formatted 
                        (year, month, day)
                        example: (2017, 3, 22)
                        
        end_date    --  The end of a time window (tuple) formatted 
                        (year, month, day)
                        example: (2017, 3, 22)
        
        timezone    --  "PST", "PDT", "MST", "MDT", "GMT"
        
        col_names   -- Optional list for df column names
        
        set_day     -- Boolean, True sets day 
                        
                        
    Returns:
        
        df -- A pandas dataframe with metadata from the webservice is returned.  
              Metadata is stored in df.__dict__['metadata'], the data is used in 
              some of the plotting functions.  The metadata is easily lost if a df
              is copied or transformed in some way.  It may be best to export the 
              metadata if it is needed.  meta = df.__dict__['metadata']
        
        
    """
    try: 
        lookback = kwargs['lookback']
        start_date = False
        end_date = False
    except:
        lookback = False
        start_date = kwargs['start_date']
        end_date = kwargs['end_date']
    try:timezone = kwargs['timezone']
    except: timezone = 'PST'
    url = time_window_url(paths,start_date=start_date, end_date=end_date, lookback = lookback, public=public,timezone = timezone)
    if public:
        r = requests.get(url)
    else:
        requests.packages.urllib3.disable_warnings() 
        r = requests.get(url, verify = False)
    json_data = json.loads(r.text)
    df_list = []
    meta = {}
    if not isinstance(paths, list):
        paths = [paths]
    if col_names:
         col_dict = {path:name for path, name in zip(paths, col_names)}
    site_dict = {}
    for path in paths:
        site = path.split('.')[0]
        try: site_dict[site].append(path)
        except KeyError:
            site_dict.update({site:[path]})
    freq_list = []        
    for site,path_list in site_dict.items():
        try:
            data = json_data[site]
        except KeyError:
            sys.stderr.write('No data for %s\n' % site)
            continue
        lat = data['coordinates']['latitude']
        long = data['coordinates']['longitude']
        tz_offset = data['tz_offset']
        tz = data['timezone']
        
        for path in path_list:
            
            vals = data['timeseries'][path.strip()]
            try:path_data = vals['values']
            except KeyError: 
                sys.stderr.write('!No data for %s\n' % path)
                paths = [x for x in paths if x != path]
                continue
            date = [val[0] for val in path_data]
            values = [val[1] for val in path_data]
            flags = [val[2] for val in path_data]
            df= pd.DataFrame({'date': date, path: values})
            df['date'] = pd.to_datetime(df['date'])
            flags = pd.DataFrame({'date': df['date'], 'flag': flags})
            flags = flags[flags['flag']>0].set_index('date')
            df.set_index('date', inplace = True)
            freq = get_frequency(path)
            freq_list.append(freq)
#            if freq and 'D' in freq and set_day:
#                df.index = [x.replace(hour = 0, minute = 0, second = 0) for x in df.index]
#                df.index.name = 'date'
            if freq and fill:
                start = df.dropna().index[0]
                end = df.dropna().index[-1]
                start_date = (start.year, start.month, start.day)
                end_date = (end.year, end.month, end.day)
                df = df.pipe(fill_index, start_date, end_date, freq)
                
            df_list.append(df)
            vals.pop('values', None)
            vals.update({'path':path, 'lat':lat,'long':long, 
                         'tz_offset':tz_offset, 'timezone':tz, 'flags': flags})
            meta.update({path:vals})
    
    if not df_list: return False
    else: df = pd.concat(df_list, axis = 1)
    if len(set(freq_list)) == 1:
        try:
            df = df.asfreq(freq_list[0])
        except ValueError:
            pass
    df =df[paths]
    if col_names:
        df.rename(columns = col_dict, inplace = True)
        new_meta = {}
        for path in col_dict.keys():
            new_meta.update({col_dict[path]:meta[path]})
        meta = new_meta
    df.__dict__['metadata'] = meta   
    return df
    

def catalog():
    
    """
    
    Requests the CWMS catalog.  Returns a large dict and not easy 
    wade through, it would be easier to go to a dataquery site to find 
    what you are looking for http://www.nwd-wc.usace.army.mil/dd/common/dataquery/www/
        
    Arguments: 
    
    Returns: dict
    
    """
    url = r'http://www.nwd-wc.usace.army.mil/dd/common/web_service/webexec/getjson?catalog=%5B%5D'
    r = requests.get(url)
    return json.loads(r.text)

def site_catalog(site):
    
    """
    Returns a dictionary of CWMS data paths for a particular site
    
    Arguments:
        site -- cwms site name, example TDDO
    
    Returns: 
        json.loads(r.text) -- dictionary of available site data
        
    """
    
    url = r'http://www.nwd-wc.usace.army.mil/dd/common/web_service/webexec/getjson?tscatalog=%5B%22SITE%22%5D'
    url = url.replace('SITE', site.upper())
    r = requests.get(url)
    return json.loads(r.text)

 

File: cwms_read.r
library(lubridate)
library(data.table)
library(magrittr)
time_parse = function(time1, time2){
  
  time = as.numeric(difftime(time2,time1))
  forward_weeks = time / 7
  forward_days = (forward_weeks - as.integer(forward_weeks)) * 7
  forward_hours = (forward_days - as.integer(forward_days)) * 24
  forward_minutes = (forward_hours - as.integer(forward_hours)) * 60
  times = c(forward_weeks,forward_days,forward_hours,forward_minutes)
  v = c()
  for (i in 1:length(times)){
    v[i] = as.integer(times[i])
    
  }
  return(v)
}



time_window_url = function(path, start_date, end_date, timezone = 'PST'){
  url = 'http://nwp-wmlocal2.nwp.usace.army.mil/common/web_service/webexec/getjson?timezone=TIMEZONE&query=%5B%22PATH%22%2C%22PATH%22%5D&startdate=START_MONTH%2FSTART_DAY%2FSTART_YEAR+00%3A00&enddate=END_MONTH%2FEND_DAY%2FEND_YEAR+00%3A00'
  
  start_date = strptime(start_date,"%Y-%m-%d")
  end_date = strptime(end_date, "%Y-%m-%d")
  now = strptime(Sys.time(),"%Y-%m-%d %H:%M:%S")
  
  
  start = time_parse(start_date, now)
  end = time_parse(end_date, now)
  s_week = paste(start[1], 'w', sep ='')
  s_day = paste(start[2], 'd', sep ='')
  s_hour = paste(start[3], 'h', sep ='')
  s_minute = paste(start[4], 'm', sep ='')
  e_week = paste(end[1], 'w', sep ='')
  e_day = paste(end[2], 'd', sep ='')
  e_hour = paste(end[3], 'h', sep ='')
  e_minute = paste(end[4], 'm', sep ='')
  start_month = toString(month(start_date))
  start_day = toString(day(start_date))
  start_year = toString(year(start_date))
  end_month = toString(month(end_date))
  end_day = toString(day(end_date))
  end_year = toString(year(end_date))
  
  
  
  url = gsub('BACKWARD_WEEK_', s_week, url) %>% gsub('BACKWARD_DAY_',s_day,.) %>% 
      gsub('BACKWARD_HOUR_', s_hour, .) %>% gsub('BACKWARD_MINUTE_',s_minute, .) %>%
      gsub('FORWARD_WEEK_', e_week, .) %>% gsub('FORWARD_DAY_',e_day, .) %>% 
      gsub('FORWARD_HOUR_', e_hour, .) %>% gsub('FORWARD_MINUTE_',e_minute, .) %>% 
      gsub('START_MONTH', start_month,.) %>% gsub('START_DAY', start_day, .) %>%
      gsub('START_YEAR', start_year, .) %>% gsub('END_MONTH', end_month,.) %>%
      gsub('END_DAY', end_day, .) %>% gsub('END_YEAR', end_year, .)  %>% 
      gsub("PATH", path, .)  %>% gsub('TIMEZONE', timezone, .)
  
  
  return(url)
  
}


cwms_data_parse = function(data, column){
  value = data[column][[1]]
  return(value) 
}


cwms_to_dt = function(path, start_date, end_date, timezone = 'PST'){
  url = time_window_url(path, start_date, end_date, timezone = timezone)
  data = rjson::fromJSON(file=url)
  paste(paste('`',path,sep=''),'`',sep='')
  x = data[[1]]$timeseries[[1]]$values
  name = strsplit(path, split = "\\.")[[1]]
  name = paste(name[1],name[2], sep = '_')
  dt = data.table(date = unlist(lapply(x,cwms_data_parse,1)), 
                  value = unlist(lapply(x,cwms_data_parse,2)))
  dt$value = as.numeric(dt$value)
  setnames(dt, c('date', name))
  dt$date = as.POSIXct(dt$date, format = "%Y-%m-%dT%H:%M:%S")
  return(dt)
}



get_cwms = function(paths, start_date, end_date, timezone = 'PST'){

  dt = lapply(paths,cwms_to_dt, start_date = start_date, end_date = end_date, timezone = timezone)
  return(Reduce(merge,dt))
}



